heron/packing/src/java/com/twitter/heron/packing/roundrobin/RoundRobinPacking.java

```diff
@@ -343,4 +353,32 @@ private void validatePackingPlan(PackingPlan plan) throws PackingException {
       }
     }
   }
+
+  @Override
+  public PackingPlan repack(PackingPlan currentPackingPlan, Map<String, Integer> componentChanges)
+      throws PackingException {
+    int initialNumContainer = TopologyUtils.getNumContainers(topology);
+    int initialNumInstance = TopologyUtils.getTotalInstance(topology);
+    double initialNumInstancePerContainer = (double) initialNumInstance / initialNumContainer;
```

Comment:
This implementation tries to keep the ratio of `#(instances)/#(containers)` always the same during re-scaling.

The issue here is:
During each re-scale, we would update only # of instances in topology def, and # of containers will not be updated. So the # of containers fetched by `TopologyUtils.getNumContainers(topology)` is not accurate. So if we continues scaling-up a topology, we would end up:
**increasing the # of instances in one single container boundlessly**.

One thought is always getting # of instances and containers based on `currentPackingPlan`. This will reflect the correct values.
But it also has some issues: we will try to keep the ratio of `#(instances)/#(containers)` the same with last `currentPackingPlan`, rather than the original one. And every time we would ceil the # of containers, since the # of containers has to be an integer. And after several times of scale-up, it can lead to:
**The # of instances in per container decreases and eventually there will be only one instance per container.**

The ideal case is to always keep the ratio of `#(instances)/#(containers)` the same as the **original** ratio. But it seems our current state updating is kind of messy, and I am not sure whether there is a simple way to get it. Or one thought is to insert this original ratio as a config during submission.

Reply:
I am curious about how the original number of container is evaluated? Is it configured or calculated? If it is configured, maybe better to pass a parameter? If it is calculated based on topology, can we run the same calculation with the new topology config to have a good number of container? I agree with maosong that the ratio may or may not be the best option here.

Reply:
- I am curious about how the original number of container is evaluated? 
It is set by customers by is put into the topology-level config: https://github.com/twitter/heron/blob/master/heron/api/src/java/com/twitter/heron/api/Config.java#L356


One issue is there is a bug when updating the topology: https://github.com/twitter/heron/blob/master/heron/scheduler-core/src/java/com/twitter/heron/scheduler/UpdateTopologyManager.java#L317. It will update only the new # of instances, but not the new # of containers.
Also, I don't think it is a good idea to update the topology def here; I think it should be unchanged once a topology is submitted. All runtime changes should apply to the PhycicalPlan.


- I agree with maosong that the ratio may or may not be the best option here.
Ratio is ok if we can keep the ratio same as the original one. 
Another thought is we could also accept `updated count of containers` when updating the topology, rather than accept `updated count of instances for component`. To make it, we can reuse existing routine: the component name is `container` or `stmgr`, like:
`--component-parallelism container:100`

Reply:
how about count the max#(instance) and min#(instance) per container from the current packing plan, and keep both counts during scaling ?

Reply:
make the ratio remains same during topology lifecycle #2602